src/node/rpc/file_serving_handlers.h

```diff
@@ -0,0 +1,466 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include "ccf/base_endpoint_registry.h"
+#include "ccf/service/tables/nodes.h"
+#include "snapshots/filenames.h"
+
+namespace ccf::node
+{
+  // Helper function to lookup redirect address based on the interface on this
+  // node which received the request. Will either return an address, or
+  // populate an appropriate error on the response context.
+  // Takes both CommandEndpointContext and ReadOnlyTx, so that it can be
+  // called be either read-only or read-write endpoints
+  static std::optional<std::string> get_redirect_address_for_node(
+    const ccf::endpoints::CommandEndpointContext& ctx,
+    ccf::kv::ReadOnlyTx& ro_tx,
+    const ccf::NodeId& target_node)
+  {
+    auto nodes = ro_tx.ro<ccf::Nodes>(ccf::Tables::NODES);
+
+    auto node_info = nodes->get(target_node);
+    if (!node_info.has_value())
+    {
+      LOG_FAIL_FMT("Node redirection error: Unknown node {}", target_node);
+      ctx.rpc_ctx->set_error(
+        HTTP_STATUS_INTERNAL_SERVER_ERROR,
+        ccf::errors::InternalError,
+        fmt::format(
+          "Cannot find node info to produce redirect response for node {}",
+          target_node));
+      return std::nullopt;
+    }
+
+    const auto interface_id = ctx.rpc_ctx->get_session_context()->interface_id;
+    if (!interface_id.has_value())
+    {
+      LOG_FAIL_FMT("Node redirection error: Non-RPC request");
+      ctx.rpc_ctx->set_error(
+        HTTP_STATUS_INTERNAL_SERVER_ERROR,
+        ccf::errors::InternalError,
+        "Cannot redirect non-RPC request");
+      return std::nullopt;
+    }
+
+    const auto& interfaces = node_info->rpc_interfaces;
+    const auto interface_it = interfaces.find(interface_id.value());
+    if (interface_it == interfaces.end())
+    {
+      LOG_FAIL_FMT(
+        "Node redirection error: Target missing interface {}",
+        interface_id.value());
+      ctx.rpc_ctx->set_error(
+        HTTP_STATUS_INTERNAL_SERVER_ERROR,
+        ccf::errors::InternalError,
+        fmt::format(
+          "Cannot redirect request. Received on RPC interface {}, which is "
+          "not present on target node {}",
+          interface_id.value(),
+          target_node));
+      return std::nullopt;
+    }
+
+    const auto& interface = interface_it->second;
+    return interface.published_address;
+  }
+
+  static void init_file_serving_handlers(
```

Comment:
To avoid making `node_frontend.h` ever larger and more complex, I've moved the snapshot endpoints here. It leads to a bit of awkwardness in accessing subsystems (hence the plumbing of a new `get_primary()` call), but I think its worthwhile for readability.